crates/simulate/src/agent/agent_channel.rs

```diff
@@ -1,53 +1,138 @@
-use std::borrow::Cow;
+use crossbeam_channel::{Receiver, Sender};
+use ethers::abi::Detokenize;
+use revm::primitives::TxEnv;
 
-use crossbeam_channel::Sender;
-
-use super::types::AgentId;
 use crate::{
     address::Address,
-    state::update::{SimUpdate, UpdateData},
+    state::{
+        update::{EvmStateUpdate, EvmStateUpdateOutput, Update, WorldStateUpdate},
+        EvmTxOutput, StateError, StateMiddleware,
+    },
+    utils::decode_output,
 };
 
-#[derive(Debug, Clone)]
-pub struct AgentSimUpdate<U: UpdateData> {
-    pub update: SimUpdate<U>,
-    pub address: Address,
-    pub tag: Option<Cow<'static, str>>,
+pub type AgentChannelResult<T> = Result<T, AgentChannelError>;
+
+#[derive(thiserror::Error, Debug)]
+pub enum AgentChannelError {
+    #[error("Error interacting with state.")]
+    StateError(#[from] StateError),
+    #[error("No more tx outputs to receive.")]
+    MissingEvmTxOutput,
+    #[error("No more tx outputs to receive.")]
+    MissingWorldUpdateOutput,
+    #[error("Tried receiving wrong tx output type.")]
+    WrongEvmTxOutputType,
+    #[error("Tried receiving tx output without tx function.")]
+    MissingEvmTxFunction,
 }
-pub struct AgentChannel<U: UpdateData> {
-    pub sender: Sender<AgentSimUpdate<U>>,
-    pub address: Address,
-    pub tag: Option<Cow<'static, str>>,
+
+pub struct AgentChannelSender<U: Update> {
+    address: Address,
+    pub evm_update_sender: Sender<EvmStateUpdate>,
+    pub evm_result_sender: Sender<EvmStateUpdateOutput>,
+    pub world_update_sender: Sender<WorldStateUpdate<U>>,
+    pub world_result_sender: Sender<U>,
 }
 
-impl<U: UpdateData> AgentChannel<U> {
-    pub fn new(sender: &Sender<AgentSimUpdate<U>>, agent_id: &AgentId) -> Self {
-        AgentChannel {
-            address: agent_id.address,
-            sender: sender.clone(),
-            tag: agent_id.clone().name.into(),
+impl<U: Update> AgentChannelSender<U> {
```

Comment:
Agent channel sender is now the way to interact with both the evm and `World`. It wraps the either a revm `TxEnv` or a ethers `ContractCall` on behalf of the agent and sends it to the manager for execution.